### PR TITLE
Use webp for avatar storage (cgo variant)

### PR DESCRIPTION
**Pull request by zeripath**
This PR uses the github.com/chai2010/webp package to encode and store avatars as webp.

Ref #24263
Close #24661

### CODE REVIEW

**Comment by silverwind** [comment]
Fails on Windows as suspected. I'm really not sure if this dependency is the way to go. We should to the remaining avatar changes as discussed and maybe wait for a better webp encoder to appear in the ecosystem.

**Comment by lunny** [comment]
I don't think we should introduce another CGO library.

**Comment by silverwind** [comment]
Agree, our goal should be to build with `CGO_ENABLED=0` which will give the best  go toolchain and produces binaries that can run standalone inside docker `from: scratch`.

**Comment by zeripath** [comment]
> I don't think we should introduce another CGO library.

Then we cannot store avatars as webp. We have four options:

1. Continue to store as png and abandon any hope of storing as webp.
2. Use a CGO that uses libwebp -  this library appears to be only one that correctly inserts the cgo dependency.
3. Use the ccgo library - https://git.sr.ht/~jackmordaunt/go-libwebp 
4. One of us needs to learn the webp spec, write pure go code to encode as webp and either upstream it or maintain it as a library

> ... maybe wait for a better webp encoder to appear in the ecosystem.

None will appear as far as I can see. The referenced issue for the improvement to x/image/png is extremely old.

> Fails on Windows as suspected.

That test is poorly named - it's actually testing if we can build with `CGO_ENABLED=0` ... 

> Agree, our goal should be to build with CGO_ENABLED=0, see https://github.com/go-gitea/gitea/issues/24658 for the topic.

I've added a build-tag switch code which will switch to png when `CGO_ENABLED=0` ... 

Whilst having everything in pure go is kinda nice I am suspicious that we may be needlessly holding things back and may be choosing inferior options simply because of it. 

> We should do the remaining avatar processing changes as discussed .... 

Certainly, but if webp is wanted - this or the ccgo library are the only options. I don't think there's anything else.

---

Anyway I'll get this to pass CI and leave it as an option.

**Comment by silverwind** [comment]
> None will appear as far as I can see. The referenced issue for the improvement to x/image/png is extremely old.

Full browser support only came [very recently](https://caniuse.com/webp), so webp adoption will increase significantly now, and I assume it will eventually spawn a usable pure-go library.

File: modules/avatar/encode_cgo.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+//go:build cgo
+
+package avatar
+
+import (
+	"image"
+	"io"
+
+	"code.gitea.io/gitea/modules/log"
+	"github.com/chai2010/webp"
```
Comment (zeripath) [inline modules/avatar/encode_cgo.go:14]:
```suggestion

	"github.com/chai2010/webp"
```

**Comment by silverwind** [comment]
I think we can take https://github.com/go-gitea/gitea/pull/24653 as a viable alternative. webp will be preserved in original if below the configured size. If above, it will resize and be stored as PNG. I think it's a good compromise until a pure-go webp encoder arises.

**Comment by wxiaoguang** [comment]
Actually, for further steps, maybe we can parse the webp header manually (just like what I have done in an early commit), then accept webp by default. I guess our "PNG resize/compression" never wins over webp.

Then we can drop the webp package totally, reduce Gitea's codebase side.

**Comment by lunny** [comment]
Since #24653 merged, maybe we can close this one at the moment?

**Comment by silverwind** [comment]
I think we agree on closing this. Maybe later when there is a better go webp encode, we can do it, and use this as a reference possibly.

**Pull request closed without merging**